Xamarin.PropertyEditing.Mac/Controls/Custom/AutocompleteComboBox.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Reflection;
+using AppKit;
+using Foundation;
+using ObjCRuntime;
+using Xamarin.PropertyEditing.ViewModels;
+
+namespace Xamarin.PropertyEditing.Mac.Controls
+{
+	internal class AutocompleteComboBox : NSComboBox
+	{
+		private readonly PropertyViewModel viewModel;
+		private readonly PropertyInfo previewCustomExpressionPropertyInfo;
+		private ObservableCollectionEx<string> values;
+
+		public AutocompleteComboBox (PropertyViewModel viewModel, ObservableCollectionEx<string> values, PropertyInfo previewCustomExpressionPropertyInfo)
+		{
+			if (viewModel == null)
+				throw new ArgumentNullException (nameof (viewModel));
+
+			if (values == null)
+				throw new ArgumentNullException (nameof (values));
+
+			if (previewCustomExpressionPropertyInfo == null)
+				throw new ArgumentNullException (nameof (previewCustomExpressionPropertyInfo));
+
+			this.viewModel = viewModel;
+			this.values = values;
+			this.previewCustomExpressionPropertyInfo = previewCustomExpressionPropertyInfo;
+
+			this.values.CollectionChanged += (sender, e) => {
+				switch (e.Action) {
+					case NotifyCollectionChangedAction.Add:
+					if (e.NewItems != null) {
+						var items = (ObservableCollectionEx<string>)e.NewItems;
+						int startIndex = e.NewStartingIndex;
+						if (startIndex != -1 && startIndex < items.Count) {
+							for (var i = 0; i < items.Count; i++) {
+								Insert (new NSString ((string)items[i]), startIndex);
+								startIndex++;
+							}
+						} else {
+							PopulateComboBoxItems (items);
+						}
+					}
+					break;
+
+					case NotifyCollectionChangedAction.Remove:
+						if (e.OldItems != null) {
+							var items = (ObservableCollectionEx<string>)e.OldItems;
+							int startIndex = e.OldStartingIndex;
+							if (startIndex != -1 && startIndex < items.Count) {
+								for (var i = 0; i < items.Count; i++) {
+									RemoveAt (startIndex);
+									startIndex++;
```

Comment:
`0 | 1 | 2 | 3 | 4 | 5`
```
startIndex = 1; count = 2;
RemoveAt (startIndex++);
```
This leaves you in a state where:
```
0 | 2 | 3 | 4 | 5
startIndex == 2;
RemoveAt (startIndex++);
```
You've removed item 1 and 3 instead of 1 and 2. Don't increment the index when removing by index.

Reply:
OK Sure. Also just noticed I might end up with an out of bounds on the **RemoveAt** too.
I also noticed that everywhere else, where we are tracking **NotifyCollectionChangedAction.Add** and **NotifyCollectionChangedAction.Remove**, that we aren't using **NewStartingIndex** or ** OldStartingIndex**. Shouldn't we be doing that across the board for consistency??